jgrapht-core/src/main/java/org/jgrapht/generate/PruferTreeGenerator.java

```diff
@@ -0,0 +1,179 @@
+     * @throws IllegalArgumentException if {@code n} is &le; 0
+     */
+    public PruferTreeGenerator(int n) {
+        this(n, new Random());
+    }
+
+    /**
+     * Construct a new PruferTreeGenerator.
+     *
+     * @param n number of vertices to be generated
+     * @param seed seed for the random number generator
+     * @throws IllegalArgumentException if {@code n} is &le; 0
+     */
+    public PruferTreeGenerator(int n, long seed) {
+        this(n, new Random(seed));
+    }
+
+    /**
+     * Construct a new PruferTreeGenerator
+     *
+     * @param n number of vertices to be generated
+     * @param rng the random number generator to use
+     * @throws IllegalArgumentException if {@code n} is &le; 0
+     * @throws NullPointerException if {@code rng} is {@code null}
+     */
+    public PruferTreeGenerator(int n, Random rng) {
+        if (n <= 0){
+            throw new IllegalArgumentException("n must be greater than 0");
+        }
+
+        this.n = n;
+        this.rng = Objects.requireNonNull(rng, "Random number generator cannot be null");
+    }
+
+    /**
+     * Generates a tree.
+     *
+     * Note: All existing vertices and edges of the target graph will be removed.
+     *
+     * @param target the target graph
+     * @param resultMap not used by this generator, can be null
+     * @throws NullPointerException if {@code target} is {@code null}
+     * @throws IllegalArgumentException if {@code target} is not undirected
+     */
+    @Override
+    public void generateGraph(Graph<V, E> target, Map<String, V> resultMap) {
+        GraphTests.requireUndirected(target);
+
+        // remove old vertices and edges
+        target.removeAllVertices(new HashSet<>(target.vertexSet()));
+
+        // base case
+        if (n == 1){
+            if (target.addVertex() == null) {
+                throw new IllegalArgumentException("Invalid vertex supplier");
+            }
+
+            return;
+        }
+
+        List<V> vertexList = new ArrayList<>(n);
+
+        // add vertices
+        for (int i = 0; i < n; i++) {
+            V newVertex = target.addVertex();
+
+            if (newVertex == null) {
+                throw new IllegalArgumentException("Invalid vertex supplier");
+            }
+
+            vertexList.add(newVertex);
+        }
+
+        // degree stores the remaining degree (plus one) for each node. The
+        // degree of a node in the decoded tree is one more than the number
+        // of times it appears in the code.
+        int[] degree = new int[n];
+        Arrays.fill(degree, 1);
+
+        int[] pruferSeq = new int[n - 2];
+        for (int i = 0; i < n - 2; i++) {
+            pruferSeq[i] = rng.nextInt(n);
+            ++degree[pruferSeq[i]];
+        }
+
+        int index = -1, x = -1;
+
+        for (int k = 0; k < n; k++){
+            if (degree[k] == 1){
+                index = x = k;
+                break;
+            }
+        }
+
+        assert index != -1;
```

Comment:
Shouldn't x's initialization be deferred to here with x = index?

Reply:
I tried to keep it identical to the implementation presented in the paper. It doesn't really matter that much.